File: app/adapters/ebay.py
"""Adapter implementation for the eBay Motors listings feed."""
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional
from urllib.parse import urlencode

from bs4 import BeautifulSoup

from app.adapters.base import BaseAdapter, ScrapedCar
from app.models import SearchCriteria


@dataclass(slots=True)
class SelectorConfig:
    """CSS selectors describing the structure of an eBay search result card."""

    listing_container: str = "li.s-item"
    title: str = ".s-item__title"
    price: str = ".s-item__price"
    mileage: str = ".s-item__dynamic"
    location: str = ".s-item__location"
    image: str = ".s-item__image-img"


class EbayAdapter(BaseAdapter):
    """Scrape eBay Motors car listings."""

    SITE_NAME = "ebay"
    BASE_URL = "https://www.ebay.com/sch/Cars-Trucks/6001"

    def __init__(self, *, session=None) -> None:
        super().__init__(self.SITE_NAME, self.BASE_URL, session=session)
        self.selectors = SelectorConfig()

    def build_search_url(self, criteria: SearchCriteria) -> str:
        params: dict[str, str] = {}
        keywords: List[str] = []
        if criteria.make:
            keywords.append(criteria.make)
        if criteria.model:
            keywords.append(criteria.model)
        if criteria.keywords:
            keywords.extend(criteria.keywords)
        if keywords:
            params["_nkw"] = " ".join(keywords)
        if criteria.min_price is not None:
            params["_udlo"] = str(int(criteria.min_price))
        if criteria.max_price is not None:
            params["_udhi"] = str(int(criteria.max_price))
        query = urlencode(params)
        return f"{self.base_url}?{query}" if query else self.base_url

    def _parse_soup(self, soup: BeautifulSoup, criteria: SearchCriteria) -> List[ScrapedCar]:
        listings: List[ScrapedCar] = []
        selector = self.selectors
        for idx, card in enumerate(soup.select(selector.listing_container), start=1):
            title = self._text(card.select_one(selector.title)) or "eBay Listing"
            price_text = self._text(card.select_one(selector.price))
            price = self._parse_price(price_text) or 0.0
            details_block = self._text(card.select_one(selector.mileage))
            mileage = self._parse_int(details_block)
            location_text = self._text(card.select_one(selector.location)) or criteria.location

            image_url: Optional[str] = None
            image_tag = card.select_one(selector.image)
            if image_tag and image_tag.has_attr("src"):
                image_url = image_tag["src"]
            elif image_tag and image_tag.has_attr("data-src"):
                image_url = image_tag["data-src"]

            listing_id = card.get("data-view") or f"ebay-{idx}"

            listings.append(
                ScrapedCar(
                    listing_id=listing_id,
                    website=self.site_name,
                    title=title,
                    price=price,
                    mileage_km=mileage,
                    location=location_text,
                    images=[image_url] if image_url else [],
                )
            )

        return listings


__all__ = ["EbayAdapter", "SelectorConfig"]

File: src/lib/ebay.ts
import * as cheerio from "cheerio";

import {
  AdapterError,
  SearchCriteria,
  ScrapedCar,
  milesToKm,
  parseMileageMiles,
  parsePrice,
} from "./types";

const SITE_NAME = "ebay";
const BASE_URL = "https://www.ebay.co.uk/sch/Cars-/9801/i.html";

function buildSearchUrl(criteria: SearchCriteria): string {
  const params = new URLSearchParams();
  const keywords: string[] = [];
  if (criteria.make) {
    keywords.push(criteria.make);
  }
  if (criteria.model) {
    keywords.push(criteria.model);
  }
  if (criteria.bodyType) {
    keywords.push(criteria.bodyType);
  }
  if (keywords.length) {
    params.set("_nkw", keywords.join(" "));
  }
  if (typeof criteria.minPrice === "number") {
    params.set("_udlo", String(Math.floor(criteria.minPrice)));
  }
  if (typeof criteria.maxPrice === "number") {
    params.set("_udhi", String(Math.floor(criteria.maxPrice)));
  }
  params.set("_sop", criteria.sort === "price-asc" ? "15" : "10"); // 10 = newly listed, 15 = price asc
  if (criteria.page && criteria.page > 1) {
    params.set("_pgn", String(criteria.page));
  }
  return `${BASE_URL}?${params.toString()}`;
}

async function fetchHtml(url: string): Promise<string> {
  const response = await fetch(url, {
    headers: {
      "user-agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 " +
        "(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    },
  });
  if (!response.ok) {
    throw new AdapterError(`eBay returned ${response.status}`);
  }
  return response.text();
}

export async function searchEbay(criteria: SearchCriteria): Promise<{
  cars: ScrapedCar[];
  sourceUrl: string;
}> {
  const url = buildSearchUrl(criteria);
  const html = await fetchHtml(url);
  const $ = cheerio.load(html);
  const cards = $("li.s-item");
  const limit = criteria.limit ?? 16;
  const cars: ScrapedCar[] = [];

  cards.each((index, element) => {
    if (cars.length >= limit) {
      return false;
    }
    const title =
      $(element).find(".s-item__title").text().replace(/\s+/g, " ").trim() ||
      "eBay Listing";
    if (title === "Shop on eBay") {
      return;
    }
    const priceText = $(element)
      .find(".s-item__price")
      .first()
      .text()
      .replace(/\s+/g, " ")
      .trim();
    const mileageText = $(element)
      .find(".s-item__dynamic")
      .first()
      .text()
      .replace(/\s+/g, " ")
      .trim();
    const mileageMiles = parseMileageMiles(mileageText);
    const location = $(element)
      .find(".s-item__location")
      .first()
      .text()
      .replace(/Item location:\s*/i, "")
      .trim();
    const imageTag = $(element).find(".s-item__image-img").first();
    const image =
      imageTag.attr("src") ?? imageTag.attr("data-src") ?? undefined;
    const link =
      $(element).find(".s-item__link").attr("href") ??
      "https://www.ebay.co.uk";
    const listingId =
      $(element).attr("data-view") ||
      link.split("/").filter(Boolean).pop() ||
      `${SITE_NAME}-${index + 1}`;

    cars.push({
      listingId,
      website: SITE_NAME,
      title,
      link,
      price: parsePrice(priceText),
      currency: "GBP",
      mileageMiles,
      mileageKm: milesToKm(mileageMiles),
      location: location || undefined,
      image,
      sellerType: mileageText.includes("Dealer") ? "trade" : undefined,
      textBlock: `${title} ${mileageText}`.trim(),
    });
  });

  return { cars, sourceUrl: url };
}
